File: algo/algo.java
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.Random;

public class Matching {
    private int groupSize;
    private int[][] prefs;
    private int iterCount;
    private int finalIterCount;
    private int numMembers;
    private int numGroups;
    private List<Integer> ungrouped;
    private List<Group> unfilled;
    private List<Group> filled;

    public Matching(int[][] prefs, int groupSize, int iterCount, int finalIterCount) {
        this.groupSize = groupSize;
        this.prefs = prefs;
        this.iterCount = iterCount;
        this.finalIterCount = finalIterCount;
        this.numMembers = prefs.length;
        this.numGroups = (int) Math.ceil((double) numMembers / groupSize);
        this.ungrouped = new ArrayList<>();
        for (int i = 0; i < numMembers; i++) {
            this.ungrouped.add(i);
        }
        this.unfilled = new ArrayList<>();
        this.filled = new ArrayList<>();
        Random random = new Random();
        for (int i : random.ints(0, numMembers).distinct().limit(numGroups).toArray()) {
            Group group = new Group(this, new ArrayList<>(Arrays.asList(i)));
            this.unfilled.add(group);
            this.ungrouped.remove((Integer) i);
        }
    }

    public static Matching fromCsv(String file_path, int r) {
        // You'll need to implement file reading and parsing in Java
        // and pass the preferences as a 2D array (prefs) to the constructor.
        int[][] prefs = {}; // Parse CSV data
        return new Matching(prefs, r, 2, 4);
    }

    public double getMemPrefForGroup(int mem, List<Integer> grp) {
        double pref = 0;
        for (int i : grp) {
            pref += prefs[mem][i];
        }
        pref = pref * (1.0 / grp.size());
        return pref;
    }

    public double getGroupPrefForMem(int mem, List<Integer> grp) {
        double pref = 0;
        for (int i : grp) {
            pref += prefs[i][mem];
        }
        pref = pref * (1.0 / grp.size());
        return pref;
    }

    public double getGroupScore(List<Integer> y) {
        if (y.size() <= 1) {
            return 0;
        }
        double score = 0;
        for (int i : y) {
            for (int j : y) {
                if (i != j) {
                    score += prefs[i][j];
                }
            }
        }
        score = score * (1.0 / (y.size() * y.size() - y.size()));
        return score;
    }

    public double getNetScore() {
        double score = 0;
        for (Group group : filled) {
            score += getGroupScore(group.getMembers());
        }
        return score / numGroups;
    }

    public List<Object> solve() {
        while (!ungrouped.isEmpty()) {
            addOneMember();
        }

        filled.addAll(unfilled);
        unfilled.clear();
        optimize(true);
        List<List<Integer>> grps = new ArrayList<>();
        for (Group group : filled) {
            grps.add(group.getMembers());
        }
        List<Object> result = new ArrayList<>();
        result.add(getNetScore());
        result.add(grps);
        return result;
    }

    public void optimize(boolean useFilled) {
        List<Group> grps = useFilled ? filled : unfilled;
        int iters = useFilled ? finalIterCount : iterCount;

        for (int a = 0; a < iters; a++) {
            for (Group grp1 : grps) {
                for (int mem1 : new ArrayList<>(grp1.getMembers())) {
                    if (mem1 == -1) {
                        break;
                    }
                    if (grp1 == grp2) {
                        continue;
                    }
                    for (Group grp2 : grps) {
                        if (mem1 == -1) {
                            break;
                        }
                        if (grp2 == grp1) {
                            continue;
                        }
                        for (int mem2 : new ArrayList<>(grp2.getMembers())) {
                            if (mem1 == -1) {
                                break;
                            }
                            if (mem2 == mem1) {
                                continue;
                            }
                            List<Integer> grp2mem1 = new ArrayList<>(grp2.getMembers());
                            grp2mem1.remove((Integer) mem2);
                            grp2mem1.add(mem1);
                            List<Integer> grp1mem2 = new ArrayList<>(grp1.getMembers());
                            grp1mem2.remove((Integer) mem1);
                            grp1mem2.add(mem2);

                            double grpOneNewScore = getGroupScore(grp1mem2);
                            double grpTwoNewScore = getGroupScore(grp2mem1);

                            if (grpOneNewScore + grpTwoNewScore > getGroupScore(grp1.getMembers()) + getGroupScore(grp2.getMembers())) {
                                grp1.addMember(mem2);
                                grp1.removeMember(mem1);
                                grp2.addMember(mem1);
                                grp2.removeMember(mem2);
                                mem1 = -1;
                            }
                        }
                    }
                }
            }
        }
    }

    public void addOneMember() {
        boolean[][] proposed = new boolean[ungrouped.size()][unfilled.size()];
        boolean[] isTempGrouped = new boolean[ungrouped.size()];
        double[][] tempPref = new double[ungrouped.size()][unfilled.size()];
        int[][] tempPrefOrder = new int[ungrouped.size()][unfilled.size()];

        for (int i = 0; i < ungrouped.size(); i++) {
            for (int j = 0; j < unfilled.size(); j++) {
                tempPref[i][j] = getMemPrefForGroup(ungrouped.get(i), unfilled.get(j).getMembers());
            }
        }

        for (int i = 0; i < ungrouped.size(); i++) {
            int[] indices = new int[unfilled.size()];
            for (int j = 0; j < unfilled.size(); j++) {
                indices[j] = j;
            }
            Arrays.sort(indices, (a, b) -> Double.compare(tempPref[i][b], tempPref[i][a]));
            tempPrefOrder[i] = indices;
        }

        while (Arrays.stream(isTempGrouped).filter(g -> !g).count() != 0) {
            for (int i = 0; i < ungrouped.size(); i++) {
                if (isTempGrouped[i]) {
                    continue;
                }
                if (Arrays.stream(proposed[i]).filter(p -> !p).count() == 0) {
                    isTempGrouped[i] = true;
                    continue;
                }
                for (int j : tempPrefOrder[i]) {
                    if (proposed[i][j]) {
                        continue;
                    }
                    Group grp = unfilled.get(j);
                    proposed[i][j] = true;
                    double pref = getGroupPrefForMem(ungrouped.get(i), grp.getMembers());
                    if (pref > grp.getTempScore()) {
                        if (grp.getTempMember() >= 0) {
                            isTempGrouped[ungrouped.indexOf(grp.getTempMember())] = false;
                        }
                        grp.addTemp(ungrouped.get(i));
                        isTempGrouped[i] = true;
                        break;
                    }
                }
            }
        }

        for (Group grp : unfilled) {
            if (grp.getTempMember() < 0) {
                continue;
            }
            ungrouped.remove((Integer) grp.getTempMember());
            grp.addPermanently();
        }

        optimize(false);

        for (Group grp : unfilled) {
            if (grp.size() >= groupSize || ungrouped.isEmpty()) {
                filled.add(grp);
            }
        }

        for (Group grp : filled) {
            unfilled.remove(grp);
        }
    }
}

class Group {
    private Matching game;
    private List<Integer> members;
    private int tempMember;
    private double tempScore;

    public Group(Matching game, List<Integer> members) {
        this.game = game;
        this.members = members;
        this.tempMember = -1;
        this.tempScore = -1;
    }

    public void addMember(int x) {
        members.add(x);
    }

    public void removeMember(int x) {
        members.remove((Integer) x);
    }

    public double addTemp(int x) {
        tempMember = x;
        tempScore = game.getGroupPrefForMem(x, members);
        return tempScore;
    }

    public void addPermanently() {
        if (tempMember == -1) {
            return;
        }
        addMember(tempMember);
        tempMember = -1;
        tempScore = -1;
    }

    public int size() {
        return members.size();
    }

    public List<Integer> getMembers() {
        return members;
    }

    public int getTempMember() {
        return tempMember;
    }

    public double getTempScore() {
        return tempScore;
    }
}

File: algo.js
class Matching {
  constructor(prefs, group_size = 4, iter_count = 2, final_iter_count = 4) {
    this.group_size = group_size;
    this.prefs = prefs;
    this.iter_count = iter_count;
    this.final_iter_count = final_iter_count;
    this.num_members = prefs.length;
    this.num_groups = Math.ceil(this.num_members / group_size);
    this.ungrouped = Array.from({ length: this.num_members }, (_, i) => i);
    this.unfilled = [];
    this.filled = [];

    const randomIndices = new Set();
    while (randomIndices.size < this.num_groups) {
      randomIndices.add(Math.floor(Math.random() * this.num_members));
    }

    for (const i of randomIndices) {
      this.unfilled.push(new Group(this, [i]));
      this.ungrouped.splice(this.ungrouped.indexOf(i), 1);
    }
  }

  static from_csv(file_path, r = 4) {
    // You'll need to implement file reading and parsing in JavaScript
    // and pass the preferences as a 2D array (prefs) to the constructor.
    const prefs = ...; // Parse CSV data
    return new Matching(prefs, r);
  }

  get_mem_pref_for_group(mem, grp) {
    let pref = 0;
    for (const i of grp) {
      pref += this.prefs[mem][i];
    }
    pref *= 1.0 / grp.length;
    return pref;
  }

  get_group_pref_for_mem(mem, grp) {
    let pref = 0;
    for (const i of grp) {
      pref += this.prefs[i][mem];
    }
    pref *= 1.0 / grp.length;
    return pref;
  }

  get_group_score(y) {
    if (y.length <= 1) {
      return 0;
    }
    let score = 0;
    for (const i of y) {
      for (const j of y) {
        if (i !== j) {
          score += this.prefs[i][j];
        }
      }
    }
    score *= 1.0 / (y.length ** 2 - y.length);
    return score;
  }

  get_net_score() {
    let score = 0;
    for (const i of this.filled) {
      score += this.get_group_score(i.members);
    }
    return score / this.num_groups;
  }

  solve() {
    while (this.ungrouped.length !== 0) {
      this.add_one_member();
    }

    this.filled.push(...this.unfilled);
    this.unfilled = [];
    this.optimize(true);
    const grps = this.filled.map((i) => i.members);
    return [this.get_net_score(), grps];
  }

  optimize(use_filled = true) {
    const grps = use_filled ? this.filled : this.unfilled;
    const iters = use_filled ? this.final_iter_count : this.iter_count;

    for (let a = 0; a < iters; a++) {
      for (const grp1 of grps) {
        for (const mem1 of grp1.members) {
          if (mem1 === -1) {
            break;
          }
          if (grp1 === grp2) {
            continue;
          }
          for (const grp2 of grps) {
            if (mem1 === -1) {
              break;
            }
            if (grp2 === grp1) {
              continue;
            }
            for (const mem2 of grp2.members) {
              if (mem1 === -1) {
                break;
              }
              if (mem2 === mem1) {
                continue;
              }
              const grp2mem1 = [...grp2.members];
              grp2mem1.splice(grp2mem1.indexOf(mem2), 1);
              grp2mem1.push(mem1);
              const grp1mem2 = [...grp1.members];
              grp1mem2.splice(grp1mem2.indexOf(mem1), 1);
              grp1mem2.push(mem2);

              const grp_one_new_score = this.get_group_score(grp1mem2);
              const grp_two_new_score = this.get_group_score(grp2mem1);

              if (grp_one_new_score + grp_two_new_score >
                this.get_group_score(grp1.members) +
                this.get_group_score(grp2.members)) {
                grp1.add_member(mem2);
                grp1.remove_member(mem1);
                grp2.add_member(mem1);
                grp2.remove_member(mem2);
                mem1 = -1;
              }
            }
          }
        }
      }
    }
  }

  add_one_member() {
    const proposed = Array.from({ length: this.ungrouped.length }, () => Array(this.unfilled.length).fill(false));
    const is_temp_grouped = new Array(this.ungrouped.length).fill(false);
    const temp_pref = Array.from({ length: this.ungrouped.length }, () => Array(this.unfilled.length).fill(0));
    const temp_pref_order = Array.from({ length: this.ungrouped.length }, () => Array(this.unfilled.length).fill(0));

    for (let i = 0; i < this.ungrouped.length; i++) {
      for (let j = 0; j < this.unfilled.length; j++) {
        temp_pref[i][j] = this.get_mem_pref_for_group(this.ungrouped[i], this.unfilled[j].members);
      }
    }

    for (let i = 0; i < this.ungrouped.length; i++) {
      temp_pref_order[i] = temp_pref[i].slice().sort((a, b) => b - a);
    }

    while (is_temp_grouped.filter((grouped) => !grouped).length !== 0) {
      for (let i = 0; i < this.ungrouped.length; i++) {
        if (is_temp_grouped[i]) {
          continue;
        }
        if (proposed[i].filter((p) => !p).length === 0) {
          is_temp_grouped[i] = true;
          continue;
        }
        for (const j of temp_pref_order[i]) {
          if (proposed[i][j]) {
            continue;
          }
          const grp = this.unfilled[j];
          proposed[i][j] = true;
          const pref = this.get_group_pref_for_mem(this.ungrouped[i], grp.members);
          if (pref > grp.tempScore) {
            if (grp.tempMember >= 0) {
              is_temp_grouped[this.ungrouped.indexOf(grp.tempMember)] = false;
            }
            grp.add_temp(this.ungrouped[i]);
            is_temp_grouped[i] = true;
            break;
          }
        }
      }
    }

    for (const grp of this.unfilled) {
      if (grp.tempMember < 0) {
        continue;
      }
      this.ungrouped.splice(this.ungrouped.indexOf(grp.tempMember), 1);
      grp.add_permanently();
    }

    this.optimize(false);

    for (const grp of this.unfilled) {
      if (grp.size() >= this.group_size || this.ungrouped.length === 0) {
        this.filled.push(grp);
      }
    }

    for (const grp of this.filled) {
      this.unfilled.splice(this.unfilled.indexOf(grp), 1);
    }
  }
}

class Group {
  constructor(game, members = []) {
    this.game = game;
    this.members = members;
    this.tempMember = -1;
    this.tempScore = -1;
  }

  add_member(x) {
    this.members.push(x);
  }

  remove_member(x) {
    const index = this.members.indexOf(x);
    if (index !== -1) {
      this.members.splice(index, 1);
    }
  }

  add_temp(x) {
    this.tempMember = x;
    this.tempScore = this.game.get_group_pref_for_mem(x, this.members);
    return this.tempScore;
  }

  add_permanently() {
    if (this.tempMember === -1) {
      return;
    }
    this.add_member(this.tempMember);
    this.tempMember = -1;
    this.tempScore = -1;
  }

  size() {
    return this.members.length;
  }
}
